src/lightning/pytorch/core/datamodule.py

```diff
@@ -244,3 +245,34 @@ def load_from_checkpoint(
             **kwargs,
         )
         return cast(Self, loaded)
+
+    def __str__(self) -> str:
+        """Return a string representation of the datasets that are setup.
+
+        Returns:
+            A string representation of the datasets that are setup.
+
+        """
+        datasets_info: Optional[list[str]] = []
+
+        def len_implemented(obj: Dataset) -> bool:
+            try:
+                len(obj)
+                return True
+            except NotImplementedError:
+                return False
+
+        for attr_name in dir(self):
+            attr = getattr(self, attr_name)
+
+            # Get Dataset information
+            if isinstance(attr, Dataset):
```

Comment:
The problem with this is that there are no guarantees that any datasets actually exist as attributes.

I think a better design would be to leverage the contract, and get dataloaders through  `self.train_dataloader` (caveat: it may be a dataloader or an iterable of dataloaders), `self.val_dataloader`, etc, access their `dataset` property (https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader) and produce a string from there in a similar way to the current way.

Also instead of ignoring non-Dataset subclasses in this case, we could call str from whatever is returned as dataset.

I think it's a valuable addition, wdyt?

Reply:
That is a good idea. I also had a similar idea in the beginning but discarded it because I thought it would add too much overhead. But I did some research an DataLoader should not add too much overhead to the actual dataset so we can use it in the str method.
Your suggestion would make the method more save, while still maintaining the same functionality, since one has to call prepare and setup anyway before one is able to access the dataset information.
I will try to implement your suggestion